src/multiselect_contextmenu.js

```diff
@@ -480,6 +595,80 @@ const registerDelete = function() {
   Blockly.ContextMenuRegistry.registry.register(deleteOption);
 };
 
+/**
+ * Paste multiple selected blocks from clipboard.
+ * @param {boolean} useCopyPasteCrossTab Whether to use cross tab copy paste.
+ */
+const registerPaste = function(useCopyPasteCrossTab) {
+  const pasteOption = {
+    displayText: function() {
+      const workableBlocksLength = blockNumGetFromStorage(useCopyPasteCrossTab);
+      if (workableBlocksLength <= 1) {
+        return Blockly.Msg['CROSS_TAB_PASTE']?
+          Blockly.Msg['CROSS_TAB_PASTE'] : 'Paste';
+      } else {
+        return Blockly.Msg['CROSS_TAB_PASTE_X_BLOCKS']?
+            Blockly.Msg['CROSS_TAB_PASTE_X_BLOCKS'].replace(
+                '%1', workableBlocksLength) :
+            (Blockly.Msg['CROSS_TAB_PASTE']?
+            Blockly.Msg['CROSS_TAB_PASTE'] : 'Paste'
+            ) + ' (' +
+            workableBlocksLength + ')';
+      }
+    },
+    preconditionFn: function(scope) {
+      if (blockNumGetFromStorage(useCopyPasteCrossTab) < 1) {
+        return 'disabled';
+      }
+      return !scope.workspace.options.readOnly?
+        'enabled': 'hidden';
```

Comment:
It seems like we should always return hidden if the workspace is read-only, but as implemented if cross tab copy paste is enabled the state will either be disabled or hidden depending on whether there is content in the local storage. If it's disabled, that might imply to users that at some point it could be enabled, which will never be the case. Update this so that the read only check is first so that the option is hidden regardless of other factors.

Reply:
For paste, now it will always be hidden when the workspace is read-only, if it's not, then it depends on the content in the clipboard to decide whether it's enabled or disabled.